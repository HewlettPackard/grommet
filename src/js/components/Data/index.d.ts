import * as React from 'react';
import { BoxProps } from '../Box/index';

export interface View {
  properties?: {
    [property: string]:
      boolean | (string | number)[] | { max: number, min: number };
  };
  search?: string | {
    property?: string | string[];
    text: string;
  };
  sort?: {
    property: string;
    direction: 'asc' | 'desc';
  };

  // Future column ordering, requires 'schema' property on Data
  // columns?: string[];

  // Future name
  // name?: string;
}

export interface DataProps {
  data: object[];
  total?: number;

  view?: View;
  onView?: (view: View) => void;

  schema?: 'raw' | string[] | {
    [key: string]: {
      // for DataTable column header, DataFilter label, DataTableColumns label
      label?: string | React.ReactNode;
      // for DataFilter options
      values?: (string | number)[];
    };
  };

  // Future selection
  // select: string[];
  // onSelect: (select: string[]) => void;

  toolbar: boolean | 'search' | 'filters';

  // whether to add filter controls and when changes should be delivered
  // updateOn?: 'change' | 'submit';

  // Future named views
  // views?: View[];
  // onCreateView?: (view: View) =>  void;
  // onDeleteView?: (string) => void;
}

type divProps = Omit<JSX.IntrinsicElements['div'], 'onClick'>;

export interface DataExtendedProps
  extends BoxProps,
  DataProps,
    divProps {}

declare const Data: React.FC<DataExtendedProps>;

export { Data };
