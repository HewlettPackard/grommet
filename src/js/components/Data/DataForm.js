import React, { useContext, useEffect, useState } from 'react';
import styled from 'styled-components';
import { Box } from '../Box';
import { Button } from '../Button';
import { Footer } from '../Footer';
import { Form } from '../Form';
import { DataContext } from '../../contexts/DataContext';
import { MessageContext } from '../../contexts/MessageContext';

const HideableButton = styled(Button)`
  ${(props) =>
    props.disabled &&
    `
  opacity: 0;`}
`;

const hideButtonProps = {
  'aria-hidden': true,
  disabled: true,
  tabIndex: -1,
};

// We convert the view structure to something more flat to work better
// with the Form inputs. These keys are how we flatten the Form value object
// from the view object.
export const formSearchKey = '_search';
export const formSortKey = '_sort';
export const formRangeKey = '_range';
export const formStepKey = '_step';
export const formPageKey = '_page';
export const formColumnsKey = '_columns';

const viewFormKeyMap = {
  search: formSearchKey,
  sort: formSortKey,
  step: formStepKey,
  page: formPageKey,
  columns: formColumnsKey,
};

// converts from the external view format to the internal Form value format
const viewToFormValue = (view) => {
  const result = { ...(view?.properties || {}) };
  // convert { min: , max: } range to [min, max] for RangeSelector
  Object.keys(result).forEach((key) => {
    if (
      typeof result[key]?.min === 'number' ||
      typeof result[key]?.max === 'number'
    ) {
      result[key] = { [formRangeKey]: [result[key].min, result[key].max] };
    }
  });

  // convert formal view keys to their form '_' prefixed counterparts
  Object.keys(viewFormKeyMap).forEach((key) => {
    if (view?.[key]) result[viewFormKeyMap[key]] = view[key];
  });
  // always have some blank search text
  if (!result[formSearchKey]) result[formSearchKey] = '';

  if (view?.columns) result[formColumnsKey] = view.columns;

  return result;
};

// converts from the internal Form value format to the external view format
const formValueToView = (value) => {
  const result = {};

  const valueCopy = { ...value };

  Object.keys(viewFormKeyMap).forEach((key) => {
    if (valueCopy[viewFormKeyMap[key]]) {
      result[key] = valueCopy[viewFormKeyMap[key]];
      delete valueCopy[viewFormKeyMap[key]];
    }
  });

  result.properties = valueCopy;

  // convert any ranges
  Object.keys(result.properties).forEach((key) => {
    if (result.properties[key][formRangeKey]) {
      result.properties[key] = {
        min: result.properties[key][formRangeKey][0],
        max: result.properties[key][formRangeKey][1],
      };
    }
  });

  return result;
};

// remove any empty arrays of property values by deleting the key for
// that property in the view properties
const clearEmpty = (properties) => {
  Object.keys(properties)
    .filter((k) => k !== formSearchKey)
    .forEach((k) => {
      if (Array.isArray(properties[k]) && properties[k].length === 0)
        // eslint-disable-next-line no-param-reassign
        delete properties[k];
    });
};

// if paging, when anything other than the page changes, reset the page to 1
const resetPage = (nextFormValue, prevFormValue) => {
  if (prevFormValue[formPageKey] && prevFormValue[formPageKey] > 1)
    // eslint-disable-next-line no-param-reassign
    nextFormValue[formPageKey] = 1;
};

const transformTouched = (touched, value) => {
  const result = {};
  Object.keys(touched).forEach((key) => {
    result[key] = value[key];
  });
  return result;
};

export const DataForm = ({
  children,
  footer,
  gap,
  onDone,
  onTouched,
  pad,
  ...rest
}) => {
  const { messages, onView, updateOn, view } = useContext(DataContext);
  const { format } = useContext(MessageContext);
  const [formValue, setFormValue] = useState(viewToFormValue(view));
  const [changed, setChanged] = useState();

  useEffect(() => setFormValue(viewToFormValue(view)), [view]);

  return (
    <Form
      {...rest}
      value={formValue}
      onSubmit={
        updateOn === 'submit'
          ? ({ value: nextValue, touched }) => {
              clearEmpty(nextValue);
              resetPage(nextValue, formValue);
              setFormValue(nextValue);
              setChanged(false);
              if (onTouched) onTouched(transformTouched(touched, nextValue));
              onView(formValueToView(nextValue));
              if (onDone) onDone();
            }
          : undefined
      }
      onChange={(nextValue, { touched }) => {
        clearEmpty(nextValue);
        resetPage(nextValue, formValue);
        setFormValue(nextValue);
        setChanged(true);
        if (updateOn === 'change') {
          if (onTouched) onTouched(transformTouched(touched, nextValue));
          onView(formValueToView(nextValue));
        }
      }}
    >
      <Box flex={false} pad={pad} gap={gap}>
        {children}
        {footer !== false && updateOn === 'submit' && (
          <Footer>
            <Button
              label={format({
                id: 'dataForm.submit',
                messages: messages?.dataForm,
              })}
              type="submit"
              primary
            />
            <HideableButton
              label={format({
                id: 'dataForm.reset',
                messages: messages?.dataForm,
              })}
              type="reset"
              onClick={() => {
                setFormValue(viewToFormValue(view));
                setChanged(false);
              }}
              {...(!changed ? hideButtonProps : {})}
            />
          </Footer>
        )}
      </Box>
    </Form>
  );
};
