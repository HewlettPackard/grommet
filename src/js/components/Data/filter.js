// TODO: share with DataTable, List, Cards, etc.
const datumValue = (datum, property) => {
  if (!property) return undefined;
  const parts = property.split('.');
  if (parts.length === 1) return datum[property];
  if (!datum[parts[0]]) return undefined;
  return datumValue(datum[parts[0]], parts.slice(1).join('.'));
};

export const filter = (data, view, properties) => {
  const searchExp =
    view?.search?.text || (typeof view?.search === 'string' && view?.search)
      ? new RegExp(view.search.text || view.search, 'i')
      : undefined;
  const searchProperty = view?.search?.property;

  const result = data.filter((datum) => {
    let matched = true;

    // check whether it matches any search
    if (searchExp) {
      const searchWith = (property) => {
        const value = datumValue(datum, property);
        if (value === undefined) return false;
        return searchExp.test(value);
      };

      if (searchProperty) {
        // we know where we want to search, look there
        if (typeof searchProperty === 'string')
          matched = searchWith(searchProperty);
        else if (Array.isArray(searchProperty))
          matched = searchProperty.some(searchWith);
      } else if (properties) {
        // look in defined properties
        matched = Object.keys(properties).some(searchWith);
      } else {
        // look in all properties
        matched = Object.keys(datum).some((property) =>
          searchExp.test(datum[property]),
        );
      }
    }

    // check whether it matches any specific values
    if (matched && view?.properties) {
      // if any properties don't match, overall not matched
      matched = !Object.keys(view.properties).some((property) => {
        // returning true means it doesn't match the filter,
        const filterValue = view.properties[property];
        const value = datum[property];
        if (Array.isArray(filterValue) && typeof filterValue[0] === 'number')
          return value < filterValue[0] || value > filterValue[1];
        if (Array.isArray(filterValue)) return !filterValue.includes(value);
        return filterValue !== value;
      });
    }

    return matched;
  });

  if (view?.sort?.property && view?.sort?.direction) {
    const { property, direction } = view.sort;
    const sortAsc = direction === 'asc';
    const before = sortAsc ? 1 : -1;
    const after = sortAsc ? -1 : 1;
    result.sort((d1, d2) => {
      const d1Val = d1[property];
      const d2Val = d2[property];
      // sort strings via locale case insensitive
      if (
        (typeof d1Val === 'string' && typeof d2Val === 'string') ||
        (typeof d1Val === 'string' && !d2Val) ||
        (typeof d2Val === 'string' && !d1Val)
      ) {
        const sortResult = (d1Val || '').localeCompare(d2Val || '', undefined, {
          sensitivity: 'base',
        });
        return sortAsc ? sortResult : -sortResult;
      }
      // numbers are easier to sort
      if (d1Val > d2Val) return before;
      if (d1Val < d2Val) return after;

      return 0;
    });
  }

  return result;
};
