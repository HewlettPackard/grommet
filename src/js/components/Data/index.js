export { Data } from './Data';
export { DataForm } from './DataForm';
