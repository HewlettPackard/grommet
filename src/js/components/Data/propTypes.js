import PropTypes from 'prop-types';

let PropType = {};
if (process.env.NODE_ENV !== 'production') {
  PropType = {
    data: PropTypes.arrayOf(PropTypes.shape({})),
  };
}
export const DataPropTypes = PropType;
