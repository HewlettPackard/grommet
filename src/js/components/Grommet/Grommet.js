import React, { forwardRef, useEffect, useMemo, useState } from 'react';
import { createGlobalStyle } from 'styled-components';

import {
  ContainerTargetContext,
  ResponsiveContext,
  ThemeContext,
} from '../../contexts';

import {
  deepMerge,
  backgroundIsDark,
  getBreakpoint,
  getDeviceBreakpoint,
  normalizeColor,
} from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';

const FullGlobalStyle = createGlobalStyle`
  body { margin: 0; }
`;

const deviceResponsive = (userAgent, theme) => {
  // log('--deviceResponsive', userAgent, theme);
  /*
   * Regexes provided for mobile and tablet detection are meant to replace
   * a full-featured specific library due to contributing a considerable size
   * into the bundle.
   *
   * User agents found https://deviceatlas.com/blog/list-of-user-agent-strings
   */
  if (userAgent) {
    if (/(tablet|ipad|playbook|silk)|(android(?!.*mobile))/i.test(userAgent)) {
      return getDeviceBreakpoint('tablet', theme);
    }
    if (/Mobile|iPhone|Android/.test(userAgent)) {
      return getDeviceBreakpoint('phone', theme);
    }
    return getDeviceBreakpoint('computer', theme);
  }
  return undefined;
};

const Grommet = forwardRef((props, ref) => {
  const {
    children,
    full,
    containerTarget = typeof document === 'object' ? document.body : undefined,
    theme: themeProp,
    ...rest
  } = props;

  const { background, dir, themeMode, userAgent } = props;

  const [stateResponsive, setResponsive] = useState();

  const theme = useMemo(() => {
    const nextTheme = deepMerge(baseTheme, themeProp || {});

    const {
      colors: { background: themeBackground },
    } = nextTheme.global;

    nextTheme.dark = (themeMode || nextTheme.defaultMode) === 'dark';
    const color = normalizeColor(background || themeBackground, nextTheme);
    nextTheme.dark = backgroundIsDark(color, nextTheme);
    nextTheme.baseBackground = background || themeBackground;
    // This allows DataTable to intelligently set the background of a pinned
    // header or footer.
    nextTheme.background = nextTheme.baseBackground;

    if (dir) {
      nextTheme.dir = dir;
    }

    return nextTheme;
  }, [background, dir, themeMode, themeProp]);

  useEffect(() => {
    const onResize = () => {
      setResponsive(getBreakpoint(window.innerWidth, theme));
    };
    window.addEventListener('resize', onResize);
    onResize();
    return () => {
      window.removeEventListener('resize', onResize);
    };
  }, [theme]);

  const responsive =
    stateResponsive ||
    deviceResponsive(userAgent, theme) ||
    theme.global.deviceBreakpoints.tablet;

  return (
    <ThemeContext.Provider value={theme}>
      <ResponsiveContext.Provider value={responsive}>
        <ContainerTargetContext.Provider value={containerTarget}>
          <StyledGrommet full={full} {...rest} ref={ref}>
            {children}
          </StyledGrommet>
          {full && <FullGlobalStyle />}
        </ContainerTargetContext.Provider>
      </ResponsiveContext.Provider>
    </ThemeContext.Provider>
  );
});

Grommet.displayName = 'Grommet';

let GrommetDoc;
if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  GrommetDoc = require('./doc').doc(Grommet);
}
const GrommetWrapper = GrommetDoc || Grommet;

export { GrommetWrapper as Grommet };
