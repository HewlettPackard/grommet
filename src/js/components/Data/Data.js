import React, { useEffect, useMemo, useState } from 'react';
import { Box } from '../Box';
import { DataFilters } from '../DataFilters';
import { DataSearch } from '../DataSearch';
import { DataSummary } from '../DataSummary';
import { Toolbar } from '../Toolbar';
import { DataContext } from '../../contexts/DataContext';
import { DataPropTypes } from './propTypes';
import { filter } from './filter';

const defaultView = {
  search: '',
  properties: {},
};

export const Data = ({
  children,
  data: dataProp,
  id = 'data',
  messages,
  onView,
  properties,
  toolbar,
  total,
  updateOn = 'submit',
  view: viewProp,
  ...rest
}) => {
  const [view, setView] = useState(viewProp);
  useEffect(() => setView(viewProp), [viewProp]);

  const data = useMemo(() => {
    if (onView) return dataProp;
    return filter(dataProp, view, properties);
  }, [dataProp, onView, properties, view]);

  // what we use for DataContext value
  const contextValue = useMemo(() => {
    const result = { id, messages, properties, updateOn, view };

    if (
      view?.search ||
      view?.sort ||
      (view?.properties && Object.keys(view.properties).length)
    ) {
      result.clearFilters = () => {
        const nextView = defaultView;
        setView(nextView);
        if (onView) onView(nextView);
      };
    }

    result.onView = (nextView) => {
      setView(nextView);
      if (onView) onView(nextView);
    };

    result.data = data;
    result.unfilteredData = dataProp;
    result.total = total !== undefined ? total : dataProp.length;

    return result;
  }, [data, dataProp, id, messages, onView, properties, total, updateOn, view]);

  let toolbarContent;
  if (toolbar) {
    toolbarContent = [
      <Toolbar key="toolbar">
        {(toolbar === true || toolbar === 'search') && <DataSearch />}
        {(toolbar === true || toolbar === 'filters') && <DataFilters drop />}
      </Toolbar>,
      <DataSummary key="summary" />,
    ];
  }

  return (
    <DataContext.Provider value={contextValue}>
      <Box id={id} flex={false} {...rest}>
        {toolbarContent}
        {children}
      </Box>
    </DataContext.Provider>
  );
};

Data.propTypes = DataPropTypes;
