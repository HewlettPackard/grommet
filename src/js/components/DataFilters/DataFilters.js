import React, { Children, useContext, useState } from 'react';
import { Filter } from 'grommet-icons/icons/Filter';
import { Box } from '../Box';
import { Button } from '../Button';
import { DataFilter } from '../DataFilter';
import { DataForm } from '../Data';
import { DropButton } from '../DropButton';
import { Header } from '../Header';
import { Heading } from '../Heading';
import { DataContext } from '../../contexts/DataContext';
import { MessageContext } from '../../contexts/MessageContext';
import { DataFiltersPropTypes } from './propTypes';

const dropProps = {
  align: { top: 'bottom', right: 'right' },
};

export const DataFilters = ({ drop, children, heading, ...rest }) => {
  const { clearFilters, data, messages, properties, view } =
    useContext(DataContext);
  const { format } = useContext(MessageContext);
  const [showContent, setShowContent] = useState();
  const controlled = drop;

  const clearControl = clearFilters && (
    <Box flex={false}>
      <Button
        label={format({
          id: 'dataFilters.clear',
          messages: messages?.dataFilters,
        })}
        onClick={clearFilters}
      />
    </Box>
  );

  let filters;
  if (Children.count(children) === 0) {
    let filtersFor;
    if (!properties && data && data.length) filtersFor = Object.keys(data[0]);
    else if (Array.isArray(properties)) filtersFor = properties;
    else if (typeof properties === 'object')
      filtersFor = Object.keys(properties);
    else filtersFor = [];
    filters = filtersFor.map((property) => (
      <DataFilter key={property} property={property} />
    ));
  }

  const content = (
    <DataForm
      pad={controlled ? 'medium' : undefined}
      gap="small"
      onDone={() => setShowContent(false)}
      {...(!controlled ? rest : {})}
    >
      {!drop && (
        <Header>
          <Heading margin="none" level={2} size="small">
            {heading ||
              format({
                id: 'dataFilters.heading',
                messages: messages?.dataFilters,
              })}
          </Heading>
          {!controlled && clearControl}
        </Header>
      )}
      {filters}
      {children}
    </DataForm>
  );

  if (!controlled) return content;

  let badge = 0;
  if (view?.properties) badge += Object.keys(view.properties).length;
  if (view?.search) badge += 1;
  if (view?.sort) badge += 1;
  if (!badge) badge = undefined;

  // drop
  const control = (
    <DropButton
      aria-label={format({
        id: 'dataFilters.open',
        messages: messages?.dataFilters,
      })}
      kind="toolbar"
      icon={<Filter />}
      dropProps={dropProps}
      dropContent={content}
      badge={badge}
      open={showContent}
      onOpen={() => setShowContent(undefined)}
      onClose={() => setShowContent(undefined)}
    />
  );

  return (
    <Box flex={false} direction="row" gap="small" {...rest}>
      {control}
      {clearControl}
    </Box>
  );
};

DataFilters.propTypes = DataFiltersPropTypes;
