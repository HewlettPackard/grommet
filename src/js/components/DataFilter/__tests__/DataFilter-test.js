import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import { Data } from '../../Data';
import { DataFilters } from '../../DataFilters';
import { Grommet } from '../../Grommet';
import { DataFilter } from '..';

const data = [{ name: 'a' }, { name: 'b' }];

describe('DataFilter', () => {
  test('renders', () => {
    const { container } = render(
      <Grommet>
        <Data data={data}>
          <DataFilters>
            <DataFilter property="name" />
          </DataFilters>
        </Data>
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
