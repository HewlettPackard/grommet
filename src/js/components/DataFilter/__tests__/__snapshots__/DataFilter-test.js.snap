// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataFilter renders 1`] = `
.c4 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c4 g {
  fill: inherit;
  stroke: inherit;
}

.c4 *:not([stroke])[fill="none"] {
  stroke-width: 0;
}

.c4 *[stroke*="#"],
.c4 *[STROKE*="#"] {
  stroke: inherit;
  fill: none;
}

.c4 *[fill-rule],
.c4 *[FILL-RULE],
.c4 *[fill*="#"],
.c4 *[FILL*="#"] {
  fill: inherit;
  stroke: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c5 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  height: 6px;
}

.c3 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  outline: none;
  border: none;
  padding: 0;
  text-align: inherit;
  line-height: 0;
  padding: 12px;
}

.c3:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c3:focus > circle,
.c3:focus > ellipse,
.c3:focus > line,
.c3:focus > path,
.c3:focus > polygon,
.c3:focus > polyline,
.c3:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c3:focus::-moz-focus-inner {
  border: 0;
}

.c3:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c3:focus:not(:focus-visible) > circle,
.c3:focus:not(:focus-visible) > ellipse,
.c3:focus:not(:focus-visible) > line,
.c3:focus:not(:focus-visible) > path,
.c3:focus:not(:focus-visible) > polygon,
.c3:focus:not(:focus-visible) > polyline,
.c3:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c3:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

@media only screen and (max-width:768px) {
  .c5 {
    height: 3px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <button
          aria-label="Open Drop"
          class="c3"
          type="button"
        >
          <svg
            aria-label="Filter"
            class="c4"
            viewBox="0 0 24 24"
          >
            <path
              d="m3 6 7 7v8h4v-8l7-7V3H3z"
              fill="none"
              stroke="#000"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <div
        class="c5"
      />
      <div
        class="c1"
      >
        2 items
      </div>
    </div>
  </div>
</div>
`;
